### DISCUSSION TITLE
Why it is not possible to add riadz vdev to a pool from which special device was removed.

### DISCUSSION THREAD

**Discussion by kolczykm** [openzfs/zfs#15729] [General]
I wonder why it is not possible to add raidz vdev to a pool from which special device have been removed.

1. I have created a pool using single vdev and single special device:
`zpool create Pool-0 scsi-SQEMU_QEMU_HARDDISK_69809-6 special scsi-SQEMU_QEMU_HARDDISK_69809-5 -O acltype=posixacl -O compression=off -O dedup=off -O dnodesize=auto -O special_small_blocks=0 -O xattr=sa -o ashift=12 -o autoreplace=off -o failmode=wait -o multihost=on -o autotrim=off -m /Pools/Pool-0 -f`

2. Next I removed the special device
`zpool remove Pool-0 scsi-SQEMU_QEMU_HARDDISK_69809-5`

3. After removal of special device finished I tried to add raidz1 to this pool but it failed.
`zpool add Pool-0 raidz1 scsi-SQEMU_QEMU_HARDDISK_69809-1 scsi-SQEMU_QEMU_HARDDISK_69
809-2 scsi-SQEMU_QEMU_HARDDISK_69809-3`
zfs returned error message:
`cannot add to 'Pool-0': invalid config; a pool with removing/removed vdevs does not support adding raidz or dRAID vdevs`

I do understand the message. I have this pool from which I removed special device hence I cannot add raidz now. What I am interested in is why specifically this is not possible. I can add other types of vdevs like singles or mirrors but not the raidz. Does anyone know why it is not possible for zfs to perform this operation.

Commands that I have presented above are from virtual machine where I performed some tests on zfs including this specific scenario.

**Comment by amotin** [comment]
I understand why RAIDZ can not be added during the removal -- space from RAIDZ should not be used by single disk removal process, but I also wonder why RAIDZ can not be added later after removal completion.
